{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Filtering and pre-processing\n",
    "\n",
    "Given the AnnData matrix created in the previous step, run some filtering and tooling steps for later processes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/.local/lib/python3.6/site-packages/numba/errors.py:105: UserWarning: Insufficiently recent colorama version found. Numba requires colorama >= 0.3.9\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.4 anndata==0.6.19 numpy==1.16.2 scipy==1.2.1 pandas==0.24.2 scikit-learn==0.20.3 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 \n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import collections\n",
    "\n",
    "import numpy as np\n",
    "import pandas\n",
    "import scanpy\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "\n",
    "results_file = './write/tgfb1-2.h5ad'  # the file that will store the analysis results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the AnnData matrix previously created. We have duplicate gene names in this matrix because of the merge step. We will only keep one of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.read_h5ad(\"./write/merged-final.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var[\"gene_names\"] = adata.var[\"gene_names-S1\"]\n",
    "del adata.var[\"gene_names-S1\"]\n",
    "del adata.var[\"gene_names-S2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7423\n"
     ]
    }
   ],
   "source": [
    "for i, gn in enumerate(adata.var.index):\n",
    "    if gn == \"ACTA2\":\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12267 × 36246 \n",
       "    obs: 'batch', 'cellular_barcode'\n",
       "    var: 'gene_names'\n",
       "    layers: 'ambiguous', 'spliced', 'unspliced'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "A1BG           False\n",
       "A1BG-AS1       False\n",
       "A1CF           False\n",
       "A2M            False\n",
       "A2M-AS1        False\n",
       "A2ML1          False\n",
       "A2ML1-AS1      False\n",
       "A2MP1          False\n",
       "A4GALT         False\n",
       "A4GNT          False\n",
       "AAAS           False\n",
       "AACS           False\n",
       "AACSP1         False\n",
       "AADAC          False\n",
       "AADACL2        False\n",
       "AADACL2-AS1    False\n",
       "AADACL3        False\n",
       "AADACL4        False\n",
       "AADACP1        False\n",
       "AADAT          False\n",
       "AAGAB          False\n",
       "AAK1           False\n",
       "AAMDC          False\n",
       "AAMP           False\n",
       "AANAT          False\n",
       "AAR2           False\n",
       "AARD           False\n",
       "AARS           False\n",
       "AARS2          False\n",
       "AARSD1         False\n",
       "               ...  \n",
       "ZSCAN30        False\n",
       "ZSCAN31        False\n",
       "ZSCAN32        False\n",
       "ZSCAN4         False\n",
       "ZSCAN5A        False\n",
       "ZSCAN5B        False\n",
       "ZSCAN5C        False\n",
       "ZSCAN5DP       False\n",
       "ZSCAN9         False\n",
       "ZSWIM1         False\n",
       "ZSWIM2         False\n",
       "ZSWIM3         False\n",
       "ZSWIM4         False\n",
       "ZSWIM5         False\n",
       "ZSWIM5P2       False\n",
       "ZSWIM6         False\n",
       "ZSWIM7         False\n",
       "ZSWIM8         False\n",
       "ZSWIM9         False\n",
       "ZUP1           False\n",
       "ZW10           False\n",
       "ZWILCH         False\n",
       "ZWINT          False\n",
       "ZXDA           False\n",
       "ZXDB           False\n",
       "ZXDC           False\n",
       "ZYG11A         False\n",
       "ZYG11B         False\n",
       "ZYX            False\n",
       "ZZEF1          False\n",
       "Name: gene_names, Length: 36246, dtype: bool"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var[\"gene_names\"] == \"ACTA2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([4.999e+03, 1.310e+02, 4.000e+01, 8.000e+00, 7.000e+00, 5.000e+00,\n",
       "        0.000e+00, 1.000e+00, 2.000e+00, 1.000e+00]),\n",
       " array([ 0. ,  1.2,  2.4,  3.6,  4.8,  6. ,  7.2,  8.4,  9.6, 10.8, 12. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEABJREFUeJzt3X+s3XV9x/Hna9TfGgty17C2riQ2GjQTSAM4lmWDCRWM5Q81GKeda9J/2IaLiYPtDzKVBbNF1GyyEOmsjokENTTKxAYwZslAijAEKusdgrQDWi2gjqirvvfH+dQcsXf33HtP7+Hez/OR3Jzv9/39nO/3/UnLffX74xxSVUiS+vNrk25AkjQZBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUysm3cD/5/jjj69169ZNug1JWlLuuuuu71XV1GzjntMBsG7dOnbt2jXpNiRpSUnyyCjjvAQkSZ0yACSpUwaAJHXKAJCkThkAktSpkQIgycNJvpXkniS7Wu24JDuT7Gmvx7Z6knw8yXSSe5OcOrSfzW38niSbj86UJEmjmMsZwO9X1clVtaGtXwLcUlXrgVvaOsCbgPXtZytwFQwCA7gMOB04DbjscGhIkhbfQi4BbQK2t+XtwAVD9U/XwO3AyiQnAOcCO6vqYFU9CewENi7g+JKkBRg1AAr4apK7kmxttVVV9VhbfhxY1ZZXA48OvXdvq81UlyRNwKifBP6dqtqX5NeBnUm+PbyxqirJWP7v8i1gtgK88pWvXNC+1l3y5XG0NGcPX3H+RI4rSXMx0hlAVe1rr/uBLzK4hv9Eu7RDe93fhu8D1g69fU2rzVR/9rGurqoNVbVhamrWr7KQJM3TrAGQ5CVJXnZ4GTgHuA/YARx+kmczcGNb3gG8uz0NdAbwdLtUdDNwTpJj283fc1pNkjQBo1wCWgV8Mcnh8f9SVV9JcidwfZItwCPA29v4m4DzgGngGeA9AFV1MMkHgTvbuA9U1cGxzUSSNCezBkBVPQS8/gj17wNnH6FewEUz7GsbsG3ubUqSxs1PAktSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6NXIAJDkmyd1JvtTWT0xyR5LpJJ9L8vxWf0Fbn27b1w3t49JWfzDJueOejCRpdHM5A7gY2D20/mHgyqp6FfAksKXVtwBPtvqVbRxJTgIuBF4LbAQ+keSYhbUvSZqvkQIgyRrgfOCTbT3AWcANbch24IK2vKmt07af3cZvAq6rqp9U1XeAaeC0cUxCkjR3o54BfBR4P/Dztv4K4KmqOtTW9wKr2/Jq4FGAtv3pNv4X9SO8R5K0yGYNgCRvBvZX1V2L0A9JtibZlWTXgQMHFuOQktSlUc4AzgTekuRh4DoGl34+BqxMsqKNWQPsa8v7gLUAbfvLge8P14/wnl+oqqurakNVbZiamprzhCRJo5k1AKrq0qpaU1XrGNzEvbWq3gncBry1DdsM3NiWd7R12vZbq6pa/cL2lNCJwHrgG2ObiSRpTlbMPmRGfwFcl+RDwN3ANa1+DfCZJNPAQQahQVXdn+R64AHgEHBRVf1sAceXJC3AnAKgqr4GfK0tP8QRnuKpqh8Db5vh/ZcDl8+1SUnS+PlJYEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tSsAZDkhUm+keQ/ktyf5K9b/cQkdySZTvK5JM9v9Re09em2fd3Qvi5t9QeTnHu0JiVJmt0oZwA/Ac6qqtcDJwMbk5wBfBi4sqpeBTwJbGnjtwBPtvqVbRxJTgIuBF4LbAQ+keSYcU5GkjS6WQOgBn7UVp/Xfgo4C7ih1bcDF7TlTW2dtv3sJGn166rqJ1X1HWAaOG0ss5AkzdlI9wCSHJPkHmA/sBP4L+CpqjrUhuwFVrfl1cCjAG3708ArhutHeI8kaZGNFABV9bOqOhlYw+Bf7a85Wg0l2ZpkV5JdBw4cOFqHkaTuzekpoKp6CrgNeAOwMsmKtmkNsK8t7wPWArTtLwe+P1w/wnuGj3F1VW2oqg1TU1NzaU+SNAejPAU0lWRlW34R8EZgN4MgeGsbthm4sS3vaOu07bdWVbX6he0poROB9cA3xjURSdLcrJh9CCcA29sTO78GXF9VX0ryAHBdkg8BdwPXtPHXAJ9JMg0cZPDkD1V1f5LrgQeAQ8BFVfWz8U5HkjSqWQOgqu4FTjlC/SGO8BRPVf0YeNsM+7ocuHzubUqSxs1PAktSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6NWsAJFmb5LYkDyS5P8nFrX5ckp1J9rTXY1s9ST6eZDrJvUlOHdrX5jZ+T5LNR29akqTZjHIGcAh4X1WdBJwBXJTkJOAS4JaqWg/c0tYB3gSsbz9bgatgEBjAZcDpwGnAZYdDQ5K0+GYNgKp6rKq+2ZZ/COwGVgObgO1t2Hbggra8Cfh0DdwOrExyAnAusLOqDlbVk8BOYONYZyNJGtmc7gEkWQecAtwBrKqqx9qmx4FVbXk18OjQ2/a22kx1SdIEjBwASV4KfB54b1X9YHhbVRVQ42goydYku5LsOnDgwDh2KUk6gpECIMnzGPzyv7aqvtDKT7RLO7TX/a2+D1g79PY1rTZT/ZdU1dVVtaGqNkxNTc1lLpKkORjlKaAA1wC7q+ojQ5t2AIef5NkM3DhUf3d7GugM4Ol2qehm4Jwkx7abv+e0miRpAlaMMOZM4F3At5Lc02p/CVwBXJ9kC/AI8Pa27SbgPGAaeAZ4D0BVHUzyQeDONu4DVXVwLLOQJM3ZrAFQVf8GZIbNZx9hfAEXzbCvbcC2uTQoSTo6/CSwJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1atYASLItyf4k9w3VjkuyM8me9npsqyfJx5NMJ7k3yalD79ncxu9JsvnoTEeSNKpRzgA+BWx8Vu0S4JaqWg/c0tYB3gSsbz9bgatgEBjAZcDpwGnAZYdDQ5I0GbMGQFV9HTj4rPImYHtb3g5cMFT/dA3cDqxMcgJwLrCzqg5W1ZPATn41VCRJi2i+9wBWVdVjbflxYFVbXg08OjRub6vNVP8VSbYm2ZVk14EDB+bZniRpNgu+CVxVBdQYejm8v6urakNVbZiamhrXbiVJzzLfAHiiXdqhve5v9X3A2qFxa1ptprokaULmGwA7gMNP8mwGbhyqv7s9DXQG8HS7VHQzcE6SY9vN33NaTZI0IStmG5Dks8DvAccn2cvgaZ4rgOuTbAEeAd7eht8EnAdMA88A7wGoqoNJPgjc2cZ9oKqefWNZkrSIZg2AqnrHDJvOPsLYAi6aYT/bgG1z6k6SdNT4SWBJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUqRWTbmA5WnfJlydy3IevOH8ix5W0NHkGIEmdWvQASLIxyYNJppNcstjHlyQNLGoAJDkG+AfgTcBJwDuSnLSYPUiSBhb7HsBpwHRVPQSQ5DpgE/DAIvexLE3q3gN4/0FaihY7AFYDjw6t7wVOX+QedBR441taep5zTwEl2Qpsbas/SvLgAnZ3PPC9hXc1cctlHjDmueTD49rTvCyXP5flMg9wLof95iiDFjsA9gFrh9bXtNovVNXVwNXjOFiSXVW1YRz7mqTlMg9wLs9Fy2Ue4FzmarGfAroTWJ/kxCTPBy4EdixyD5IkFvkMoKoOJfkT4GbgGGBbVd2/mD1IkgYW/R5AVd0E3LRIhxvLpaTngOUyD3Auz0XLZR7gXOYkVXW0jyFJeg7yqyAkqVPLMgCWy9dNJFmb5LYkDyS5P8nFk+5pIZIck+TuJF+adC8LkWRlkhuSfDvJ7iRvmHRP85Xkz9vfrfuSfDbJCyfd06iSbEuyP8l9Q7XjkuxMsqe9HjvJHkc1w1z+tv0duzfJF5OsHPdxl10ALLOvmzgEvK+qTgLOAC5awnMBuBjYPekmxuBjwFeq6jXA61mic0qyGvgzYENVvY7BgxkXTrarOfkUsPFZtUuAW6pqPXBLW18KPsWvzmUn8Lqq+i3gP4FLx33QZRcADH3dRFX9FDj8dRNLTlU9VlXfbMs/ZPCLZvVku5qfJGuA84FPTrqXhUjycuB3gWsAquqnVfXUZLtakBXAi5KsAF4M/PeE+xlZVX0dOPis8iZge1veDlywqE3N05HmUlVfrapDbfV2Bp+bGqvlGABH+rqJJflLc1iSdcApwB2T7WTePgq8H/j5pBtZoBOBA8A/tctZn0zykkk3NR9VtQ/4O+C7wGPA01X11cl2tWCrquqxtvw4sGqSzYzRHwP/Ou6dLscAWHaSvBT4PPDeqvrBpPuZqyRvBvZX1V2T7mUMVgCnAldV1SnA/7B0LjP8knZ9fBODUPsN4CVJ/nCyXY1PDR5xXPKPOSb5KwaXg68d976XYwDM+nUTS0mS5zH45X9tVX1h0v3M05nAW5I8zOCS3FlJ/nmyLc3bXmBvVR0+E7uBQSAsRX8AfKeqDlTV/wJfAH57wj0t1BNJTgBor/sn3M+CJPkj4M3AO+soPLO/HANg2XzdRJIwuNa8u6o+Mul+5quqLq2qNVW1jsGfx61VtST/pVlVjwOPJnl1K53N0v068+8CZyR5cfu7djZL9Ib2kB3A5ra8Gbhxgr0sSJKNDC6bvqWqnjkax1h2AdBumhz+uondwPVL+OsmzgTexeBfzPe0n/Mm3ZT4U+DaJPcCJwN/M+F+5qWdxdwAfBP4FoPfB0vmk7RJPgv8O/DqJHuTbAGuAN6YZA+DM5wrJtnjqGaYy98DLwN2tv/2/3Hsx/WTwJLUp2V3BiBJGo0BIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSp/4PTWo4H2ATm9cAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(adata[:,adata.var[\"gene_names\"] == \"ACTA2\"][adata.obs[\"batch\"] == \"S1\",:].X.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([6.602e+03, 2.580e+02, 1.070e+02, 8.300e+01, 1.200e+01, 4.000e+00,\n",
       "        3.000e+00, 2.000e+00, 1.000e+00, 1.000e+00]),\n",
       " array([ 0. ,  1.3,  2.6,  3.9,  5.2,  6.5,  7.8,  9.1, 10.4, 11.7, 13. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEZ9JREFUeJzt3X+s3XV9x/HnS+qvOWfLuGtYW1cSGw0uU0gDOJdls7MUNJY/1GCcdK5J/2HOLUscbH+Q+WPBbJnTbLIQqBbHRIIaGmXiTcWYJQMpwlCorncoazugdxbYJlGHvvfH+ZSd4b2757S39/T283wkN+fzfX8/53ve3+b2vu73xzk3VYUkqT/PmXQDkqTJMAAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnVox6Qb+P2eccUatX79+0m1I0rJyzz33/HtVTS0076QOgPXr17N3795JtyFJy0qSh0eZ5ykgSeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnq1En9TuDjtf6Kz0/kdb9z9Rsm8rqSNA6PACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSp0YKgCQrk9yS5JtJ9iV5TZLTk0wn2d8eV7W5SfKRJDNJ7k9y7tB2trX5+5NsO1E7JUla2KhHAB8GvlBVrwBeBewDrgD2VNUGYE9bBrgI2NC+dgDXACQ5HbgKOB84D7jqaGhIkpbeggGQ5CXArwLXA1TVD6vqCWArsKtN2wVc0sZbgRtq4E5gZZIzgQuB6ao6UlWPA9PAlkXdG0nSyEY5AjgLmAU+luTeJNcleRGwuqoeaXMeBVa38RrgwNDzD7bafHVJ0gSMEgArgHOBa6rqHOB7/O/pHgCqqoBajIaS7EiyN8ne2dnZxdikJGkOowTAQeBgVd3Vlm9hEAiPtVM7tMfDbf0hYN3Q89e22nz1/6Oqrq2qjVW1cWpqapx9kSSNYcEAqKpHgQNJXt5Km4AHgd3A0Tt5tgG3tvFu4LJ2N9AFwJPtVNHtwOYkq9rF382tJkmagBUjznsXcGOS5wEPAe9kEB43J9kOPAy8tc29DbgYmAGeanOpqiNJ3gfc3ea9t6qOLMpeSJLGNlIAVNV9wMY5Vm2aY24Bl8+znZ3AznEalCSdGL4TWJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdWqkAEjynSRfT3Jfkr2tdnqS6ST72+OqVk+SjySZSXJ/knOHtrOtzd+fZNuJ2SVJ0ijGOQL49ap6dVVtbMtXAHuqagOwpy0DXARsaF87gGtgEBjAVcD5wHnAVUdDQ5K09I7nFNBWYFcb7wIuGarfUAN3AiuTnAlcCExX1ZGqehyYBrYcx+tLko7DqAFQwBeT3JNkR6utrqpH2vhRYHUbrwEODD33YKvNV5ckTcCKEef9SlUdSvJzwHSSbw6vrKpKUovRUAuYHQAvfelLF2OTkqQ5jHQEUFWH2uNh4LMMzuE/1k7t0B4Pt+mHgHVDT1/bavPVn/1a11bVxqraODU1Nd7eSJJGtmAAJHlRkhcfHQObgW8Au4Gjd/JsA25t493AZe1uoAuAJ9upotuBzUlWtYu/m1tNkjQBo5wCWg18NsnR+X9XVV9Icjdwc5LtwMPAW9v824CLgRngKeCdAFV1JMn7gLvbvPdW1ZFF2xNJ0lgWDICqegh41Rz17wKb5qgXcPk829oJ7By/TUnSYvOdwJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1auQASHJaknuTfK4tn5XkriQzST6V5Hmt/vy2PNPWrx/axpWt/q0kFy72zkiSRjfOEcC7gX1Dyx8EPlRVLwMeB7a3+nbg8Vb/UJtHkrOBS4FXAluAjyY57fjalyQdq5ECIMla4A3AdW05wOuAW9qUXcAlbby1LdPWb2rztwI3VdUPqurbwAxw3mLshCRpfKMeAfwl8B7gx235Z4EnqurptnwQWNPGa4ADAG39k23+M/U5nvOMJDuS7E2yd3Z2doxdkSSNY8EASPJG4HBV3bME/VBV11bVxqraODU1tRQvKUldWjHCnNcCb0pyMfAC4GeADwMrk6xov+WvBQ61+YeAdcDBJCuAlwDfHaofNfwcSdISW/AIoKqurKq1VbWewUXcL1XV24E7gDe3aduAW9t4d1umrf9SVVWrX9ruEjoL2AB8ddH2RJI0llGOAObzh8BNSd4P3Atc3+rXA59IMgMcYRAaVNUDSW4GHgSeBi6vqh8dx+tLko7DWAFQVV8GvtzGDzHHXTxV9X3gLfM8/wPAB8ZtUpK0+HwnsCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tSCAZDkBUm+muSfkjyQ5E9a/awkdyWZSfKpJM9r9ee35Zm2fv3Qtq5s9W8lufBE7ZQkaWGjHAH8AHhdVb0KeDWwJckFwAeBD1XVy4DHge1t/nbg8Vb/UJtHkrOBS4FXAluAjyY5bTF3RpI0ugUDoAb+qy0+t30V8DrgllbfBVzSxlvbMm39piRp9Zuq6gdV9W1gBjhvUfZCkjS2ka4BJDktyX3AYWAa+Bfgiap6uk05CKxp4zXAAYC2/kngZ4frczxHkrTERgqAqvpRVb0aWMvgt/ZXnKiGkuxIsjfJ3tnZ2RP1MpLUvbHuAqqqJ4A7gNcAK5OsaKvWAofa+BCwDqCtfwnw3eH6HM8Zfo1rq2pjVW2cmpoapz1J0hhGuQtoKsnKNn4h8HpgH4MgeHObtg24tY13t2Xa+i9VVbX6pe0uobOADcBXF2tHJEnjWbHwFM4EdrU7dp4D3FxVn0vyIHBTkvcD9wLXt/nXA59IMgMcYXDnD1X1QJKbgQeBp4HLq+pHi7s7kqRRLRgAVXU/cM4c9YeY4y6eqvo+8JZ5tvUB4APjtylJWmy+E1iSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHVqwQBIsi7JHUkeTPJAkne3+ulJppPsb4+rWj1JPpJkJsn9Sc4d2ta2Nn9/km0nbrckSQsZ5QjgaeAPqups4ALg8iRnA1cAe6pqA7CnLQNcBGxoXzuAa2AQGMBVwPnAecBVR0NDkrT0FgyAqnqkqr7Wxv8J7APWAFuBXW3aLuCSNt4K3FADdwIrk5wJXAhMV9WRqnocmAa2LOreSJJGNtY1gCTrgXOAu4DVVfVIW/UosLqN1wAHhp52sNXmq0uSJmDkAEjy08Cngd+rqv8YXldVBdRiNJRkR5K9SfbOzs4uxiYlSXMYKQCSPJfBD/8bq+ozrfxYO7VDezzc6oeAdUNPX9tq89X/j6q6tqo2VtXGqampcfZFkjSGUe4CCnA9sK+q/mJo1W7g6J0824Bbh+qXtbuBLgCebKeKbgc2J1nVLv5ubjVJ0gSsGGHOa4F3AF9Pcl+r/RFwNXBzku3Aw8Bb27rbgIuBGeAp4J0AVXUkyfuAu9u891bVkUXZC0nS2BYMgKr6ByDzrN40x/wCLp9nWzuBneM0KEk6MXwnsCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdWjAAkuxMcjjJN4ZqpyeZTrK/Pa5q9ST5SJKZJPcnOXfoOdva/P1Jtp2Y3ZEkjWqUI4CPA1ueVbsC2FNVG4A9bRngImBD+9oBXAODwACuAs4HzgOuOhoakqTJWDAAquorwJFnlbcCu9p4F3DJUP2GGrgTWJnkTOBCYLqqjlTV48A0PxkqkqQldKzXAFZX1SNt/Ciwuo3XAAeG5h1stfnqPyHJjiR7k+ydnZ09xvYkSQs57ovAVVVALUIvR7d3bVVtrKqNU1NTi7VZSdKzHGsAPNZO7dAeD7f6IWDd0Ly1rTZfXZI0IccaALuBo3fybANuHapf1u4GugB4sp0quh3YnGRVu/i7udUkSROyYqEJST4J/BpwRpKDDO7muRq4Ocl24GHgrW36bcDFwAzwFPBOgKo6kuR9wN1t3nur6tkXliVJS2jBAKiqt82zatMccwu4fJ7t7AR2jtWdJOmE8Z3AktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6teCfhNT41l/x+Ym87neufsNEXlfS8uQRgCR1ygCQpE4ZAJLUqSW/BpBkC/Bh4DTguqq6eql7OFVN6toDeP1BWo6WNACSnAb8NfB64CBwd5LdVfXgUvahxeeFb2n5WepTQOcBM1X1UFX9ELgJ2LrEPUiSWPpTQGuAA0PLB4Hzl7gHnUI87SUdu5PufQBJdgA72uJ/JfnWcWzuDODfj7+rJbdc+4aOes8HT2An4+vm3/0kcjL3/QujTFrqADgErBtaXttqz6iqa4FrF+PFkuytqo2Lsa2ltFz7BnufFHtfesu172FLfQ3gbmBDkrOSPA+4FNi9xD1IkljiI4CqejrJ7wC3M7gNdGdVPbCUPUiSBpb8GkBV3QbctkQvtyinkiZgufYN9j4p9r70lmvfz0hVTboHSdIE+FEQktSpUzIAkmxJ8q0kM0mumHQ/o0qyLskdSR5M8kCSd0+6p3ElOS3JvUk+N+lexpFkZZJbknwzyb4kr5l0T6NI8vvte+UbST6Z5AWT7mk+SXYmOZzkG0O105NMJ9nfHldNssf5zNP7n7Xvl/uTfDbJykn2eCxOuQAY+riJi4CzgbclOXuyXY3saeAPqups4ALg8mXU+1HvBvZNuolj8GHgC1X1CuBVLIN9SLIG+F1gY1X9IoMbKy6dbFf/r48DW55VuwLYU1UbgD1t+WT0cX6y92ngF6vql4B/Bq5c6qaO1ykXACzjj5uoqkeq6mtt/J8MfgitmWxXo0uyFngDcN2kexlHkpcAvwpcD1BVP6yqJybb1chWAC9MsgL4KeDfJtzPvKrqK8CRZ5W3ArvaeBdwyZI2NaK5eq+qL1bV023xTgbva1pWTsUAmOvjJpbND9GjkqwHzgHummwnY/lL4D3AjyfdyJjOAmaBj7XTV9cledGkm1pIVR0C/hz4V+AR4Mmq+uJkuxrb6qp6pI0fBVZPspnj8NvA30+6iXGdigGw7CX5aeDTwO9V1X9Mup9RJHkjcLiq7pl0L8dgBXAucE1VnQN8j5P3VMQz2vnyrQwC7OeBFyX5zcl2dexqcEvisrstMckfMzh9e+OkexnXqRgAC37cxMksyXMZ/PC/sao+M+l+xvBa4E1JvsPgtNvrkvztZFsa2UHgYFUdPdq6hUEgnOx+A/h2Vc1W1X8DnwF+ecI9jeuxJGcCtMfDE+5nLEl+C3gj8PZahvfUn4oBsGw/biJJGJyH3ldVfzHpfsZRVVdW1dqqWs/g3/xLVbUsfhutqkeBA0le3kqbgOXwNyr+FbggyU+1751NLIOL18+yG9jWxtuAWyfYy1jaH7d6D/Cmqnpq0v0ci1MuANpFmaMfN7EPuHkZfdzEa4F3MPjt+b72dfGkm+rEu4Abk9wPvBr40wn3s6B2xHIL8DXg6wz+P5+0705N8kngH4GXJzmYZDtwNfD6JPsZHNGclH8hcJ7e/wp4MTDd/q/+zUSbPAa+E1iSOnXKHQFIkkZjAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1Kn/ARiU4FvodS0sAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(adata[:,adata.var[\"gene_names\"] == \"ACTA2\"][adata.obs[\"batch\"] == \"S2\",:].X.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Shape does not fit.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-65acfce95f7b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m7423\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/matplotlib/pyplot.py\u001b[0m in \u001b[0;36mhist\u001b[0;34m(x, bins, range, density, weights, cumulative, bottom, histtype, align, orientation, rwidth, log, color, label, stacked, normed, data, **kwargs)\u001b[0m\n\u001b[1;32m   2657\u001b[0m         \u001b[0malign\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0malign\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morientation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morientation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrwidth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrwidth\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlog\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2658\u001b[0m         \u001b[0mcolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstacked\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormed\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnormed\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2659\u001b[0;31m         **({\"data\": data} if data is not None else {}), **kwargs)\n\u001b[0m\u001b[1;32m   2660\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2661\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/matplotlib/__init__.py\u001b[0m in \u001b[0;36minner\u001b[0;34m(ax, data, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1808\u001b[0m                         \u001b[0;34m\"the Matplotlib list!)\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlabel_namer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1809\u001b[0m                         RuntimeWarning, stacklevel=2)\n\u001b[0;32m-> 1810\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1811\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1812\u001b[0m         inner.__doc__ = _add_data_doc(inner.__doc__,\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/matplotlib/axes/_axes.py\u001b[0m in \u001b[0;36mhist\u001b[0;34m(self, x, bins, range, density, weights, cumulative, bottom, histtype, align, orientation, rwidth, log, color, label, stacked, normed, **kwargs)\u001b[0m\n\u001b[1;32m   6589\u001b[0m             \u001b[0;31m# this will automatically overwrite bins,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6590\u001b[0m             \u001b[0;31m# so that each histogram uses the same bins\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6591\u001b[0;31m             \u001b[0mm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhistogram\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mweights\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mw\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mhist_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6592\u001b[0m             \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# causes problems later if it's an int\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6593\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmlast\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/numpy/lib/histograms.py\u001b[0m in \u001b[0;36mhistogram\u001b[0;34m(a, bins, range, normed, weights, density)\u001b[0m\n\u001b[1;32m    778\u001b[0m     \u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mweights\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_ravel_and_check_weights\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mweights\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    779\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 780\u001b[0;31m     \u001b[0mbin_edges\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muniform_bins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_bin_edges\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mweights\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    781\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    782\u001b[0m     \u001b[0;31m# Histogram is an integer or a float array depending on the weights.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/numpy/lib/histograms.py\u001b[0m in \u001b[0;36m_get_bin_edges\u001b[0;34m(a, bins, range, weights)\u001b[0m\n\u001b[1;32m    437\u001b[0m         bin_edges = np.linspace(\n\u001b[1;32m    438\u001b[0m             \u001b[0mfirst_edge\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlast_edge\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_equal_bins\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m             endpoint=True, dtype=bin_type)\n\u001b[0m\u001b[1;32m    440\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mbin_edges\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mfirst_edge\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlast_edge\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_equal_bins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/numpy/core/function_base.py\u001b[0m in \u001b[0;36mlinspace\u001b[0;34m(start, stop, num, endpoint, retstep, dtype, axis)\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mendpoint\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mnum\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 171\u001b[0;31m         \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstop\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    172\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    173\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/base.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, idx, value)\u001b[0m\n\u001b[1;32m    382\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    383\u001b[0m             \u001b[0madata_view\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattr_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_view_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m             \u001b[0m_init_actual_AnnData\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_view\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m             \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_view\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattr_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    386\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/base.py\u001b[0m in \u001b[0;36m_init_actual_AnnData\u001b[0;34m(adata_view)\u001b[0m\n\u001b[1;32m    373\u001b[0m             \u001b[0;34m'but need a copy of the subset.\\n\\n'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m             'Call `adata_subset = adata[index].copy(filename=...)`.')\n\u001b[0;32m--> 375\u001b[0;31m     \u001b[0madata_view\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_init_as_actual\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata_view\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    376\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    377\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/base.py\u001b[0m in \u001b[0;36mcopy\u001b[0;34m(self, filename)\u001b[0m\n\u001b[1;32m   1566\u001b[0m                            \u001b[0mraw\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raw\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raw\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1567\u001b[0m                            \u001b[0mlayers\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1568\u001b[0;31m                            dtype=self._X.dtype.name if self._X is not None else 'float32')\n\u001b[0m\u001b[1;32m   1569\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1570\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mfilename\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/base.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, X, obs, var, uns, obsm, varm, layers, raw, dtype, shape, filename, filemode, asview, oidx, vidx)\u001b[0m\n\u001b[1;32m    670\u001b[0m                 \u001b[0mlayers\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m                 \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshape\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 672\u001b[0;31m                 filename=filename, filemode=filemode)\n\u001b[0m\u001b[1;32m    673\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    674\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_init_as_view\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0madata_ref\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'AnnData'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moidx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvidx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/base.py\u001b[0m in \u001b[0;36m_init_as_actual\u001b[0;34m(self, X, obs, var, uns, obsm, varm, raw, layers, dtype, shape, filename, filemode)\u001b[0m\n\u001b[1;32m    895\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    896\u001b[0m         \u001b[0;31m# layers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 897\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_layers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAnnDataLayers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlayers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    898\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    899\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__sizeof__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/anndata/layers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, adata, layers, dtype, adata_ref, oidx, vidx)\u001b[0m\n\u001b[1;32m     33\u001b[0m                 )\n\u001b[1;32m     34\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlayer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_adata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Shape does not fit.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlayer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_layers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlayer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape does not fit."
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD8CAYAAAB0IB+mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAADU9JREFUeJzt3GGI5Hd9x/H3xztTaYym9FaQu9Ok9NJ42ELSJU0Raoq2XPLg7oFF7iBYJXhgGylVhBRLlPjIhloQrtWTilXQGH0gC57cA40ExAu3ITV4FyLb03oXhawxzZOgMe23D2bSna53mX92Z3cv+32/4GD+//ntzJcfe++dndmZVBWSpO3vFVs9gCRpcxh8SWrC4EtSEwZfkpow+JLUhMGXpCamBj/JZ5M8meT7l7g+ST6ZZCnJo0lunP2YkqT1GvII/3PAgRe5/lZg3/jfUeBf1j+WJGnWpga/qh4Efv4iSw4Bn6+RU8DVSV4/qwElSbOxcwa3sRs4P3F8YXzup6sXJjnK6LcArrzyyj+8/vrrZ3D3ktTHww8//LOqmlvL184i+INV1XHgOMD8/HwtLi5u5t1L0stekv9c69fO4q90ngD2ThzvGZ+TJF1GZhH8BeBd47/WuRl4pqp+7ekcSdLWmvqUTpIvAbcAu5JcAD4CvBKgqj4FnABuA5aAZ4H3bNSwkqS1mxr8qjoy5foC/npmE0mSNoTvtJWkJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJamJQcFPciDJ40mWktx1kevfkOSBJI8keTTJbbMfVZK0HlODn2QHcAy4FdgPHEmyf9Wyvwfur6obgMPAP896UEnS+gx5hH8TsFRV56rqOeA+4NCqNQW8Znz5tcBPZjeiJGkWhgR/N3B+4vjC+NykjwK3J7kAnADef7EbSnI0yWKSxeXl5TWMK0laq1m9aHsE+FxV7QFuA76Q5Nduu6qOV9V8Vc3Pzc3N6K4lSUMMCf4TwN6J4z3jc5PuAO4HqKrvAq8Cds1iQEnSbAwJ/mlgX5Jrk1zB6EXZhVVrfgy8DSDJmxgF3+dsJOkyMjX4VfU8cCdwEniM0V/jnElyT5KD42UfBN6b5HvAl4B3V1Vt1NCSpJdu55BFVXWC0Yuxk+funrh8FnjLbEeTJM2S77SVpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDUxKPhJDiR5PMlSkrsuseadSc4mOZPki7MdU5K0XjunLUiyAzgG/BlwATidZKGqzk6s2Qf8HfCWqno6yes2amBJ0toMeYR/E7BUVeeq6jngPuDQqjXvBY5V1dMAVfXkbMeUJK3XkODvBs5PHF8Yn5t0HXBdku8kOZXkwMVuKMnRJItJFpeXl9c2sSRpTWb1ou1OYB9wC3AE+EySq1cvqqrjVTVfVfNzc3MzumtJ0hBDgv8EsHfieM/43KQLwEJV/aqqfgj8gNEPAEnSZWJI8E8D+5Jcm+QK4DCwsGrN1xg9uifJLkZP8Zyb4ZySpHWaGvyqeh64EzgJPAbcX1VnktyT5OB42UngqSRngQeAD1XVUxs1tCTppUtVbckdz8/P1+Li4pbctyS9XCV5uKrm1/K1vtNWkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgYFP8mBJI8nWUpy14use0eSSjI/uxElSbMwNfhJdgDHgFuB/cCRJPsvsu4q4G+Ah2Y9pCRp/YY8wr8JWKqqc1X1HHAfcOgi6z4GfBz4xQznkyTNyJDg7wbOTxxfGJ/7P0luBPZW1ddf7IaSHE2ymGRxeXn5JQ8rSVq7db9om+QVwCeAD05bW1XHq2q+qubn5ubWe9eSpJdgSPCfAPZOHO8Zn3vBVcCbgW8n+RFwM7DgC7eSdHkZEvzTwL4k1ya5AjgMLLxwZVU9U1W7quqaqroGOAUcrKrFDZlYkrQmU4NfVc8DdwIngceA+6vqTJJ7khzc6AElSbOxc8iiqjoBnFh17u5LrL1l/WNJkmbNd9pKUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpoYFPwkB5I8nmQpyV0Xuf4DSc4meTTJN5O8cfajSpLWY2rwk+wAjgG3AvuBI0n2r1r2CDBfVX8AfBX4h1kPKklanyGP8G8ClqrqXFU9B9wHHJpcUFUPVNWz48NTwJ7ZjilJWq8hwd8NnJ84vjA+dyl3AN+42BVJjiZZTLK4vLw8fEpJ0rrN9EXbJLcD88C9F7u+qo5X1XxVzc/Nzc3yriVJU+wcsOYJYO/E8Z7xuf8nyduBDwNvrapfzmY8SdKsDHmEfxrYl+TaJFcAh4GFyQVJbgA+DRysqidnP6Ykab2mBr+qngfuBE4CjwH3V9WZJPckOThedi/wauArSf49ycIlbk6StEWGPKVDVZ0ATqw6d/fE5bfPeC5J0oz5TltJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaGBT8JAeSPJ5kKcldF7n+N5J8eXz9Q0mumfWgkqT1mRr8JDuAY8CtwH7gSJL9q5bdATxdVb8L/BPw8VkPKklanyGP8G8ClqrqXFU9B9wHHFq15hDwb+PLXwXeliSzG1OStF47B6zZDZyfOL4A/NGl1lTV80meAX4b+NnkoiRHgaPjw18m+f5aht6GdrFqrxpzL1a4FyvcixW/t9YvHBL8mamq48BxgCSLVTW/mfd/uXIvVrgXK9yLFe7FiiSLa/3aIU/pPAHsnTjeMz530TVJdgKvBZ5a61CSpNkbEvzTwL4k1ya5AjgMLKxaswD85fjyXwDfqqqa3ZiSpPWa+pTO+Dn5O4GTwA7gs1V1Jsk9wGJVLQD/CnwhyRLwc0Y/FKY5vo65txv3YoV7scK9WOFerFjzXsQH4pLUg++0laQmDL4kNbHhwfdjGVYM2IsPJDmb5NEk30zyxq2YczNM24uJde9IUkm27Z/kDdmLJO8cf2+cSfLFzZ5xswz4P/KGJA8keWT8/+S2rZhzoyX5bJInL/VepYx8crxPjya5cdANV9WG/WP0Iu9/AL8DXAF8D9i/as1fAZ8aXz4MfHkjZ9qqfwP34k+B3xxffl/nvRivuwp4EDgFzG/13Fv4fbEPeAT4rfHx67Z67i3ci+PA+8aX9wM/2uq5N2gv/gS4Efj+Ja6/DfgGEOBm4KEht7vRj/D9WIYVU/eiqh6oqmfHh6cYvedhOxryfQHwMUafy/SLzRxukw3Zi/cCx6rqaYCqenKTZ9wsQ/aigNeML78W+MkmzrdpqupBRn/xeCmHgM/XyCng6iSvn3a7Gx38i30sw+5Lramq54EXPpZhuxmyF5PuYPQTfDuauhfjX1H3VtXXN3OwLTDk++I64Lok30lyKsmBTZtucw3Zi48Ctye5AJwA3r85o112XmpPgE3+aAUNk+R2YB5461bPshWSvAL4BPDuLR7lcrGT0dM6tzD6re/BJL9fVf+1pVNtjSPA56rqH5P8MaP3/7y5qv5nqwd7OdjoR/h+LMOKIXtBkrcDHwYOVtUvN2m2zTZtL64C3gx8O8mPGD1HubBNX7gd8n1xAVioql9V1Q+BHzD6AbDdDNmLO4D7Aarqu8CrGH2wWjeDerLaRgffj2VYMXUvktwAfJpR7Lfr87QwZS+q6pmq2lVV11TVNYxezzhYVWv+0KjL2JD/I19j9OieJLsYPcVzbjOH3CRD9uLHwNsAkryJUfCXN3XKy8MC8K7xX+vcDDxTVT+d9kUb+pRObdzHMrzsDNyLe4FXA18Zv27946o6uGVDb5CBe9HCwL04Cfx5krPAfwMfqqpt91vwwL34IPCZJH/L6AXcd2/HB4hJvsToh/yu8esVHwFeCVBVn2L0+sVtwBLwLPCeQbe7DfdKknQRvtNWkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJauJ/Acz2XLpusNoKAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(adata[:, 7423].X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying total count\n",
    "\n",
    "From this data set, we will compute the total number of RNA molecules found in each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['n_counts'] = np.ravel(adata.X.sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['n_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['n_counts'], jitter=0.4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing most expressed genes\n",
    "\n",
    "Create a plot showing the genes that are found the most across our data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata, n_top=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering by gene and cell count\n",
    "\n",
    "As a first filtering step, we filter out cells that have fewer than 300 expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now filter out genes that are expressed in fewer than 3 cells. Before doing so, we want to save a version of the complete list of genes for future analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_genes = list(adata.var[\"gene_names\"])\n",
    "sc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After this step, we have gotten additional information, including `n_genes`, automatically computing the number of genes found to be expressed in each cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying ratio of mitochondrial genes\n",
    "\n",
    "We want to compute a ratio of mitochondrial genes to total count for each cell. This is important for quality control.\n",
    "\n",
    "Citing from \"Simple Single Cell\" workflows [(Lun, McCarthy & Marioni, 2017)](https://master.bioconductor.org/packages/release/workflows/html/simpleSingleCell.html#examining-gene-level-metrics):\n",
    "\n",
    "> High proportions are indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), possibly because of loss of cytoplasmic RNA from perforated cells. The reasoning is that mitochondria are larger than individual transcript molecules and less likely to escape through tears in the cell membrane."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_genes = adata.var_names.str.startswith('MT-')\n",
    "adata.obs['percent_mito'] = np.ravel(np.sum(\n",
    "    adata[:, mito_genes].X, axis=1)) / np.ravel(np.sum(adata.X, axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get number of mito genes in our gene set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([i for i in mito_genes if i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get number of mito gene counts per cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sum(adata[:, mito_genes].X, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show violin plot of mito gene count ratio."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['percent_mito'],\n",
    "             jitter=0.4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After these analyses, we have managed to compute the following metrics for our cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[:15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering by ratio of mitochondrial genes and total count\n",
    "\n",
    "Filter out cells whose mitochondrial gene ratio or total count is too high based on these violin plots."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scatter plot and matrix before filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.scatter(adata, x='n_counts', y='percent_mito')\n",
    "sc.pl.scatter(adata, x='n_counts', y='n_genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove cells that have total counts >100000 or mitochondrial gene ratio >0.05. These are good numbers per the violin plots above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[adata.obs['n_counts'] < 100000, :]\n",
    "adata = adata[adata.obs['percent_mito'] < 0.05, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scatter plot, violin plot and matrix after filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.scatter(adata, x='n_counts', y='percent_mito')\n",
    "sc.pl.scatter(adata, x='n_counts', y='n_genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['n_counts', 'percent_mito'],\n",
    "             jitter=0.4, multi_panel=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize total count\n",
    "\n",
    "Now that we have filtered out outliers, we can normalize the count matrix by total count to reduce this source of variability. We will correct for the library size of the data matrix by rescaling it to 10000 reads per cell. Let us look at the unnormalized counts and perform the normalization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.sum(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us inspect the new counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.sum(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Log total counts\n",
    "\n",
    "Prior knowledge of microarray and bulkRNAseq means we know that a lot of effects are linear in the log of the total count, so it is a good idea to log the data now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.sum(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying and filtering highly variable genes.\n",
    "\n",
    "An important filtering metric is dispersion. Highly variable genes can be identified by this metric and filtered as important for future processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=4, min_disp=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of highly variable genes compared with non-highly variable genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "collections.Counter(adata.var[\"highly_variable\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's save our genes list before filtering out non highly variable genes and do the actual filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "before_highly_variable_genes = list(adata.var[\"gene_names\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, adata.var['highly_variable']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[\"genes_original\"] = original_genes\n",
    "adata.uns[\"genes_before_highly_variable\"] = before_highly_variable_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scale data set\n",
    "\n",
    "Gene variance alone could be a source of noise during analysis. We will scale the data set to achieve unit variance and zero mean, clipping values exceeding std 10."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.sum(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save filtered matrix to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(results_file)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
